src/hotspot/share/gc/shared/objectCountClosure.hpp

```diff
@@ -0,0 +1,24 @@
+#ifndef SHARE_GC_SHARED_OBJECTCOUNTCLOSURE_HPP
+#define SHARE_GC_SHARED_OBJECTCOUNTCLOSURE_HPP
+
+#include "gc/shared/gcId.hpp"
+#include "gc/shared/objectCountEventSender.hpp"
+#include "jfr/jfrEvents.hpp"
+#include "memory/heapInspection.hpp"
+#include "utilities/macros.hpp"
+#include "utilities/ticks.hpp"
+
+class KlassInfoEntry;
+class Klass;
+
+class ObjectCountClosure : public ObjectCountEventSender {
+    static KlassInfoTable cit;
```

Comment:
Beware the races in updating this table's entries from multiple worker threads.

Consider the two different approaches we discussed:
1. atomic updates, vs
2. thread-local updates followed by a merge of local tables at time of emission.

Consider relative performance of the two approaches especially as we scale: (a) size of the heap (and number of object types/classes) (b) the number of worker threads

Reply:
For (1) atomic updates, see class Atomic. I'd imagine this might perform worse.

For (2) thread-local updates followed by merge of local, look at class ParHeapInspectionTask for the general idea. There's definitely a footprint impact. You can allocate the thread local table in the marking loop for each worker thread and one in the heap at the start of marking in the control thread. You then destruct the contents of each as you do currently.

[Optional: I'd also consider a lighter weight object counting event that does not gather any klass histogram data, but records just the total size. Much like Jaroslav's implementation except he was getting it during collection set construction but you could piggyback it onto the marking closure like you do here.]